### DISCUSSION TITLE
What I have made: "lamp" effect on any LED using digital output in co-working mode (ie multi-threading emulation)

### DISCUSSION THREAD

**Discussion by HO44-PROJECT** [bxparks/AceRoutine#59] [Show and tell]
Hello, I wanted to thank you very much for your work. This was exactly what I wanted for my railway project.

https://github.com/HO44-PROJECT/coroutine_led_lamp

Regards

**Comment by bxparks** [comment]
Glad that you found a use for AceRoutine.

I took a quick look at your code, and found a couple of interesting things:

1) The [gradual dimming and brightening loop](https://github.com/HO44-PROJECT/coroutine_led_lamp/blob/8d275f316fbf13194d1232587417bcaa752b6644/src/coroutine_led_lamp_fx.cpp#L51) is quite nice, because it changes the PWM modulation at every iteration, which I gather produces an extremely smooth ramping up and down in light intensity.

2) I'm always entertained by the creative ways that people use C++11 lambda expressions. In [your case](https://github.com/HO44-PROJECT/coroutine_led_lamp/blob/8d275f316fbf13194d1232587417bcaa752b6644/src/coroutine_led_lamp_fx_demo.cpp#L23), you use them as function pointers to define whether the LED should turn on or off. Your solution is perfectly understandable (for someone who knows about C++ lambda expressions). But I think I would have done it differently. I would have created a single member function (e.g. `writeLED()`) in your `Coroutine_Led_Lamp_Fx` class to hold the  `digitalWrite()` call. Then pass the `GREEN_PIN` and `RED_PIN` values into each instance of the coroutine through the constructor of `Coroutine_Led_Lamp_Fx`. Then instead of two function pointers `TRIGGER_PIN _from, _to`, I would use just a single member variable, `private int _target_output_state` to hold the alternating `LOW` and `HIGH` states of the LED. Definitely a more low-brow, and unsophisticated compared to your solution, but to me, more efficient (memory and CPU) and straightforward.

But anyway, nice work on the demo!

**Reply by HO44-PROJECT** [reply]
Thank you very much for your review. You are right about the second point except that my goal is to drive multi-state signals. A departure DB signal has 6 LEDs with 5 states. The Lambdas are here to prepare this. I will have to change the lambdas dynamically regarding the final state. I should have explained this. 

By the way your library was exactly what I was looking for and extremely helpful.

**Reply by bxparks** [reply]
Got it, more complexity may be easier with lambdas.
Good luck with the project!
